Fix instant file name path retrieval in Hudi Active Timeline

atezs82 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

When submitting even a very simple query for certain types of Hudi tables (stored on S3) we were constantly receiving the following error:
```
SQL Error [65536]: Query failed (#20230710_093123_00000_bqjz2): No scheme for file system location: path/to/hudi/table/within/bucket/.hoodie
```

Stacktrace taken from the Trino coordinator:
```
...
Caused by: java.lang.IllegalArgumentException: No scheme for file system location: path/to/hudi/table/within/bucket/.hoodie
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:218)
	at io.trino.filesystem.Location.of(Location.java:74)
	at io.trino.plugin.hudi.timeline.HudiActiveTimeline.getInstantFileNamePath(HudiActiveTimeline.java:88)
	at io.trino.plugin.hudi.timeline.HudiActiveTimeline.getInstantDetails(HudiActiveTimeline.java:71)
...
```

The issue could be reproduced for this table always, on the other hand, other tables were working just fine. We were using a simple SQL query `select * from test_catalog.test_schema.table limit 1;`, but more complex queries were also failing due to this error.

This PR proposes a solution for the problem by slightly modifying the way Trino handles the Hudi timeline internally. Albeit the solution is simple, I have also added some unit test coverage about the fix. Please let me know your thoughts about it (I needed to include a Java mock library into Maven, this might not be desirable from overall project perspective). 

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

At some point during its operation, the Hudi connector tries to update ("reset") its internal state of the Hudi filesystem if "replaced instants" (files) are present on the Hudi timeline. This operation involves accessing the Hudi table's metadata folder on the object store. When this path segment is created, the corresponding `io.trino.filesystem.Location` class is created in a manner that it did not contain the scheme, just the path (see the code change below). This is forbidden in the code of that class (since it has to adhere to the format `scheme://[userInfo@]host[:port][/path]`), therefore the mentioned exception was thrown (refer to the Javadoc of the `Location` class for more information).

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:
- Fixed reading Hudi tables with replaced instants.

ebyhr commented on plugin/trino-hudi/pom.xml:
```diff
@@ -374,6 +374,13 @@
 
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
```
We basically don't use mockito in tests. Please fix `BaseHudiConnectorTest` or `TestHudiSparkCompatibility` (or both).

atezs82 replied:
The test in `trino-hudi` is ready. I have created a simple copy of the already existing test table `stock_ticks_cow` in which the Hudi timeline has `replacecommit` actions, that is present in the patchset. I have confirmed that without the code change the test is also failing exactly with the above mentioned exception. Please let me know what do you think about this solution/direction.

Still need some time to familiarize myself with the product tests and add a test case there as well for this fix.

atezs82 replied:
I extended `TestHudiSparkCompatibility` in the latest patchset, verifying that it also fails with the `No scheme for file system location:` exception above. Modified the setup method as well to have a default value for `S3_BUCKET` if that is not specified externally. Please let me know what do you think about this solution in overall.

Signing of the CLA is underway, I will send it in the upcoming days.

atezs82 replied:
@ebyhr Are you OK with the proposed changes? Thanks for taking a look again at this!

atezs82 commented:
Hey @ebyhr! Thanks for looking at this change! I'm ready to work on this more if needed - what do you think about the latest state of the PR? I have already sent the signed CLA on the 13th of July already.

ebyhr commented on plugin/trino-hudi/src/test/resources/hudi-testing-data/stock_ticks_cow_clustered/.hoodie/20230712105030944.commit:
```diff
@@ -0,0 +1,32 @@
+{
```
Please add README.md `stock_ticks_cow_clustered` directory. The file should contain steps to reproduce for creating the table, e.g. CREATE TABLE and INSERT statement.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -44,7 +43,7 @@ public class TestHudiSparkCompatibility
     public void setUp()
     {
-        bucketName = requireNonNull(System.getenv("S3_BUCKET"), "Environment variable not set: S3_BUCKET");
+        bucketName = System.getenv().getOrDefault("S3_BUCKET", "test-bucket");
```
Please revert the unrelated change.

atezs82 replied:
I did remove this from here. Do you think it would make sense to include it in a different PR though? (Eg. `TestHiveRedirectionToHudi` has it.) Without this, the environment variable shall be set to the value which is generated by the test framework if my understanding is correct.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +319,21 @@ public void testTimelineTableRedirect()
+        createNonPartitionedTableWithReplaceCommit(tableName, COW_TABLE_TYPE);
+
+        try {
+            onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName);
```
Please verify the SELECT result.

atezs82 replied:
Done - what I want to see here though is that the query passes. If it does, then replace commits were parsed in the timeline successfully. Please let me know if you thought about some additional verification here.

ebyhr replied:
Please verify the rows using `containsOnly` instead of `hasAnyRows`.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -343,6 +357,31 @@ private void createNonPartitionedTable(String tableName, String tableType)
         onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'a1', 20, 1000), (2, 'a2', 40, 2000)");
     }
 
+    private void createNonPartitionedTableWithReplaceCommit(String tableName, String tableType)
```
Please inline this method.

atezs82 replied:
Done.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/timeline/HudiActiveTimeline.java:
```diff
@@ -85,7 +85,8 @@ public Optional<byte[]> readCompactionPlanAsBytes(HudiInstant instant)
     private Location getInstantFileNamePath(String fileName)
     {
-        return Location.of(fileName.contains(SCHEMA_COMMIT_ACTION) ? metaClient.getSchemaFolderName() : metaClient.getMetaPath().path()).appendPath(fileName);
+        Location schemaFolder = metaClient.getMetaPath().appendPath(HudiTableMetaClient.SCHEMA_FOLDER_NAME);
+        return (fileName.contains(SCHEMA_COMMIT_ACTION) ? schemaFolder : metaClient.getMetaPath()).appendPath(fileName);
```
There's no build `schemaFolder` object when `fileName` doesn't contain `SCHEMA_COMMIT_ACTION`. This can be something like:
```java
    private Location getInstantFileNamePath(String fileName)
    {
        Location metaPath = metaClient.getMetaPath();
        if (fileName.contains(SCHEMA_COMMIT_ACTION)) {
            return metaPath.appendPath(HudiTableMetaClient.SCHEMA_FOLDER_NAME);
        }
        return metaPath.appendPath(fileName);
    }
```

atezs82 replied:
Thanks for the observation, I have changed the logic accordingly.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -343,6 +357,31 @@ private void createNonPartitionedTable(String tableName, String tableType)
+                          price int,
+                          ts bigint)
+                        USING hudi
+                        TBLPROPERTIES (
+                          type = '%s',
+                          primaryKey = 'id',
+                          preCombineField = 'ts',
+                          hoodie.clustering.inline = 'true',
+                          hoodie.clustering.inline.max.commits = '1')
```
Are all table properties necessary for testing replace commits?

atezs82 replied:
They should work without these, so I did remove everything not related to clustering.

atezs82 commented:
@ebyhr Thanks for all the comments and suggestions! I tried to answer them accordingly, please let me know what do you think. I have also verified that tests were failing if the code change in `HudiActiveTimeline` was not in place.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/timeline/HudiActiveTimeline.java:
```diff
@@ -85,7 +85,11 @@ public Optional<byte[]> readCompactionPlanAsBytes(HudiInstant instant)
-        return Location.of(fileName.contains(SCHEMA_COMMIT_ACTION) ? metaClient.getSchemaFolderName() : metaClient.getMetaPath().path()).appendPath(fileName);
+        Location metaPath = metaClient.getMetaPath();
+        if (fileName.contains(SCHEMA_COMMIT_ACTION)) {
+            return metaPath.appendPath(HudiTableMetaClient.SCHEMA_FOLDER_NAME);
```
Is `.appendPath(fileName)` missing? Do existing tests cover this condition?

atezs82 replied:
I have added the `appendPath(fileName)` to the schema commit path of the condition, it was clearly missing from there. 

In order to test this, we need a second table with at least one `schemacommit` action in the Hudi timeline. Will work on preparing that as well in this patchset.

atezs82 replied:
Please note my last comment below regarding this.

ebyhr commented on plugin/trino-hudi/src/test/resources/hudi-testing-data/stock_ticks_cow_clustered/README.md:
```diff
@@ -0,0 +1,74 @@
+# Context
```
Thanks for adding this README file! Is it possible to create the table with SQL? We usually use SQL when creating the resource likes `plugin/trino-delta-lake/src/test/resources`.

atezs82 replied:
Do you mean Spark SQL instead of this minimal Spark pipeline?

ebyhr replied:
Yes.

atezs82 replied:
Changed the `README.md` file accordingly. When the `schemacommit` test is ready, I will also add a `README.md` accordingly.

atezs82 replied:
Please note my last comment below regarding this.

atezs82 commented:
Regarding the comment
> Please verify the rows using containsOnly instead of hasAnyRows.
The code has been changed accordingly, but the test table contains 4 rows instead of 2. Will analyze the root cause of that as well.

atezs82 commented:
So after spending some time creating a Hudi test table with a `schemacommit` type Hudi action in it and creating a test case around it in `TestHudiSmokeTest` I came to the realization that the relevant branch in `HudiActiveTimeline::getInstantFileNamePath` is not entered (I verified this with creating a strictly temporary logger in that class and logging from that method). That method is only seem to be used when `replacecommit` actions are present in the timeline.

Nevertheless, I put up the test table to this PR, I can most certainly create a Spark SQL README on how to create it if that is needed or remove it (it has some value for sure, but it does not seem to be relevant for this issue). @ebyhr what do you think about this?

(In the meantime, I'm still looking into the test table row multiplication issue I outlined above.)

electrum commented:
This is adding a large number of test resource files -- it more than doubles the number that currently exist. Can you simplify the Hudi table files to the minimum number that reproduce the issue?

electrum commented:
Or if that is too difficult, we can skip testing it in `TestHudiSmokeTest` if this is covered by the product tests.

atezs82 commented:
@electrum Thanks for looking at this! I will check if merging the two tables is a possibility. For now the product tests only cover the `replacecommit` scenario, in order to test schema commits as well there additional changes might be needed there (as described in https://hudi.apache.org/docs/schema_evolution/), hopefully that can be also done with Spark SQL. And that does not mean that the code enters the branch in question (see above), so I wonder if we need that test at all (it certainly adds some complexity and files even though the test tables are merged).

atezs82 commented:
One more addition: it bugged me so I quickly looked at the option of altering the needed configuration for the `schemacommit` tests from product tests. Turns out that eg. `SET spark.sql.extensions=org.apache.spark.sql.hudi.HoodieSparkSessionExtension` results in the exception `org.apache.spark.sql.AnalysisException: Cannot modify the value of a static config: spark.sql.extensions`, so this might not be that easy to do from there (also it would be not desirable to modify a commonly used test SparkContext from a single test like this).

atezs82 commented:
After spending some time on analyzing the code further about the schema commit support in `HudiActiveTimeline` I suspect that this feature is not yet in use. The only real place from where `HudiActiveTimeline::getInstantDetails` is called from is `HudiTableFileSystemView::resetFileGroupsReplaced`, but even there, right before the method is called, replace commits are filtered out right before when `HudiDefaultTimeline::getCompletedReplaceTimeline` is called - so `getInstantDetails` is called only with this commit action type.

In theory future users of the Hudi timeline interface might benefit from this feature (get proper filenames for schema commits as well, which reside under the `.schema` subfolder within the Hudi metadata folder), but for now it seems it is not used for any feature the connector exposes. I have also checked and seen that this was never called from any unit tests of `trino-hudi`. 

Given this, I would suggest to remove the schema tests altogether, and fall back to just the replace commit tests, which clearly reproduced the original issue. That would make the test system simpler, we can just use the product tests if needed to test this (so that a large number of files do not need to be added to `trino-hudi`), and if we want to test in `trino-hudi` we do not need complicated Spark SQL, we can use the same method which is currently in use in the product tests.

@electrum @ebyhr What is your opinion about this?

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -27,14 +28,15 @@
-    // for ser/deser
-    public HudiReplaceCommitMetadata()
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public HudiReplaceCommitMetadata(@JsonProperty("partitionToReplaceFileIds") final Map<String, List<String>> partitionToReplaceFileIds,
```
Move the 1st argument to next line and remove final. 
```java
    public HudiReplaceCommitMetadata(
            @JsonProperty("partitionToReplaceFileIds") Map<String, List<String>> partitionToReplaceFileIds,
            @JsonProperty("compacted") Boolean compacted)
    {
```

ebyhr replied:
Could you add a commit body to explain why this change is needed to "Fix replace commit handling in Hudi Active Timeline"?

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -27,14 +28,15 @@
     {
-        partitionToReplaceFileIds = ImmutableMap.of();
-        compacted = false;
+        this.partitionToReplaceFileIds = partitionToReplaceFileIds;
+        this.compacted = compacted;
```
Are these fields nulalble?

atezs82 replied:
They are not. If eg. `partitionToReplaceFileIds` become null this code will throw an NPE (from `HudiActiveTimeline::reetFileGroupsReplaced`): 
```scala
                        // get replace instant mapping for each partition, fileId
                        return replaceMetadata.getPartitionToReplaceFileIds().entrySet().stream()
                                .flatMap(entry -> entry.getValue().stream().map(fileId ->
                                        Map.entry(new HudiFileGroupId(entry.getKey(), fileId), instant)));
```

ebyhr replied:
Then, it would be better to add `ImmutableMap.copyOf` & `requireNonNull` and change `Boolean` to `boolean`
```
        this.partitionToReplaceFileIds = ImmutableMap.copyOf(requireNonNull(partitionToReplaceFileIds, "partitionToReplaceFileIds is null"));
```

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/table/HudiTableFileSystemView.java:
```diff
@@ -222,7 +222,7 @@ private void resetFileGroupsReplaced(HudiTimeline timeline)
                     }
                     catch (IOException e) {
-                        throw new TrinoException(HUDI_BAD_DATA, "error reading commit metadata for " + instant);
+                        throw new TrinoException(HUDI_BAD_DATA, "error reading commit metadata for " + instant, e);
```
This change looks unrelated to "Fix replace commit handling in Hudi Active Timeline". Please separate a commit.

atezs82 replied:
Removed this small change, will prepare a very simple commit for this in the future.

ebyhr commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/testing/ResourceHudiTablesInitializer.java:
```diff
@@ -155,6 +155,7 @@ private static void copyDir(Path srcDir, Path dstDir)
     public enum TestingTable
     {
         HUDI_NON_PART_COW(nonPartitionRegularColumns()),
+        HUDI_NON_PART_COW_CL_SC(schemaChangedNonPartitionedRegularColumns()),
```
What's the meaning of CL and SC?

atezs82 replied:
CL means "clustered", while SC means "schema change" in this context.

ebyhr replied:
I don't think code readers can understand the meaning. Please avoid abbreviations. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-abbreviations

atezs82 replied:
Ok, I will remove the schema change tests altogether as requested anyways, and will change this as well in the next commit.

atezs82 replied:
I have removed the abbreviation, the table is called `HUDI_NON_PART_COW_CLUSTERING` (where `COW` is a well-known Hudi concept).

ebyhr commented on plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/testing/ResourceHudiTablesInitializer.java:
```diff
@@ -221,6 +222,13 @@ private static List<Column> nonPartitionRegularColumns()
 
+        private static List<Column> schemaChangedNonPartitionedRegularColumns()
+        {
+            List<Column> result = new ArrayList<>(nonPartitionRegularColumns());
+            result.add(column("description", HIVE_STRING));
+            return ImmutableList.copyOf(result);
```
No need to use ArrayList:
```java
            return ImmutableList.<Column>builder()
                    .addAll(nonPartitionRegularColumns())
                    .add(column("description", HIVE_STRING))
                    .build();
```

atezs82 replied:
Corrected that.

ebyhr commented on plugin/trino-hudi/src/test/resources/hudi-testing-data/hudi_non_part_cow_cl_sc/.hoodie/.schema/20230815074720567.schemacommit:
```diff
@@ -0,0 +1 @@
+{"schemas":[{"max_column_id":13,"version_id":20230815074720567,"type":"record","fields":[{"id":0,"name":"_hoodie_commit_time","optional":true,"type":"string","doc":""},{"id":1,"name":"_hoodie_commit_seqno","optional":true,"type":"string","doc":""},{"id":2,"name":"_hoodie_record_key","optional":true,"type":"string","doc":""},{"id":3,"name":"_hoodie_partition_path","optional":true,"type":"string"," ...
```
> https://github.com/trinodb/trino/pull/18213#issuecomment-1682783292

Let's remove these file and use product test instead.

atezs82 replied:
I have removed the tests for the schema commits.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +320,44 @@ public void testTimelineTableRedirect()
+
+        try {
+            assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName + ";"))
+                    .containsOnly(ImmutableList.of(
```
nit: `ImmutableList.of` is redundant.

atezs82 replied:
Done

codope commented:
@ebyhr @mosabua This PR is critical and affects Hudi table with clustering. I see most comments are addressed. Any chance of getting this merged sooner and released in the next version?

atezs82 commented:
Please let me know if any other things need to be done here. We are using a patched Trino image, but we would also would like to upgrade to an official Trino release having this fix soon. Thanks in advance!

codope reviewed (approved):
The change look good. Thanks @atezs82 for fixing the issue. 
Please also include me as a reviewer for any change in Hudi connector.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +320,44 @@ public void testTimelineTableRedirect()
+                        CREATE TABLE default.%s (
+                          id bigint,
+                          name string,
+                          price int,
```
`price` column looks unused. Can we remove it?

atezs82 replied:
Yes, I have removed it.

ebyhr commented on plugin/trino-hudi/src/test/resources/hudi-testing-data/hudi_non_part_cow_clustered/README.md:
````diff
@@ -0,0 +1,57 @@
+(after replacing the appropriate locations for the existing table and the new table):
+
+```sql
+CREATE TABLE hudi_non_part_cow
+USING hudi
+OPTIONS (
+hoodie.table.name = 'hudi_non_part_cow',
+hoodie.datasource.write.recordkey.field = 'rowId',
+hoodie.datasource.write.precombine.field = 'preComb',
+hoodie.table.type = 'COPY_ON_WRITE'
+)
+LOCATION 'file:///<<< EXISTING TABLE LOCATION >>>'
+CREATE TABLE hudi_non_part_cow_clustered
+USING hudi
+OPTIONS (
+   hoodie.table.name = 'hudi_non_part_cow_clustered',
+   hoodie.datasource.write.recordkey.field = 'rowId',
+   hoodie.datasource.write.precombine.field = 'preComb',
+   hoodie.table.type = 'COPY_ON_WRITE',
+   hoodie.clustering.inline = 'true',
+   hoodie.clustering.inline.max.commits = '1'
+)
+LOCATION 'file:///<<< NEW TABLE LOCATION >>>'
+AS SELECT * FROM hudi_non_part_cow
````
So, why do we need this resource files instead of product tests? I think we can simply execute this CTAS statement in product tests?

atezs82 replied:
This was needed so that this special behavior (replace commit handling) is also tested in the Hudi connector, but I also agree on that this is still a bit more complex than it should be. I removed this test altogether from the Hudi connector and let the product tests test this behavior as well in the last changeset.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +320,44 @@ public void testTimelineTableRedirect()
+                bucketName,
+                tableName));
+
+        onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'a1', 20, 1000), (2, 'a2', 40, 2000)");
```
Move to inside of `try`. Otherwise, the table won't be dropped if this statement failed for some reason.

atezs82 replied:
Ok, done.

ebyhr commented:
The commit body is too long. Please follow https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

atezs82 commented:
Thanks for the review and the suggestions!

I have shortened/slightly reformatted the commit message to contain the changes requested in https://github.com/trinodb/trino/pull/18213#discussion_r1306769114, but still have a shorter subject line (with a longer body explaining the exact changes done). Please let me know if that is acceptable.

Regarding https://github.com/trinodb/trino/pull/18213#discussion_r1326641651, I have implemented the requested changes, on the other hand, changing the `compacted` field to `boolean` from `Boolean` also had some implications on the overriden `equals` and `hashCode` methods within the `HudiReplaceCommitMetadata` class. I have modified those methods to contain the `partitionToReplaceFileIds` field, since it was missing from both methods. Please let me know if this change is acceptable here or shall we place it in an other PR.

ebyhr commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiSparkCompatibility.java:
```diff
@@ -320,6 +320,42 @@ public void testTimelineTableRedirect()
+
+        try {
+            onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'a1', 1000), (2, 'a2', 2000)");
+            assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName + ";"))
```
Remove redundant `+ ";"`.

atezs82 replied:
Ok.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -13,28 +13,34 @@
 
-    // for ser/deser
-    public HudiReplaceCommitMetadata()
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
```
Why do we need to specify `PROPERTIES` mode? Worth leaving a code comment.

atezs82 replied:
It turns out that

> Note that this mode is currently (2.5) always used for multiple-argument creators; the only ambiguous case is that of a single-argument creator.

(from https://fasterxml.github.io/jackson-annotations/javadoc/2.9/com/fasterxml/jackson/annotation/JsonCreator.Mode.html#DEFAULT), so I have removed the annotation.

atezs82 replied:
It turned out that the annotation was needed, but the specific configuration was not - I added back the annotation.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -54,13 +60,13 @@ public boolean equals(Object o)
         HudiReplaceCommitMetadata that = (HudiReplaceCommitMetadata) o;
 
-        return compacted.equals(that.compacted);
+        return compacted == that.compacted && partitionToReplaceFileIds.equals(that.partitionToReplaceFileIds);
```
Replace the order of compacted and partitionToReplaceFileIds. Also, wrap after `&&`

atezs82 replied:
Done.

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -54,13 +60,13 @@ public boolean equals(Object o)
     public int hashCode()
     {
-        return compacted.hashCode();
+        return Objects.hash(compacted, partitionToReplaceFileIds);
```
Replace the order of compacted and partitionToReplaceFileIds

ebyhr commented on plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/model/HudiReplaceCommitMetadata.java:
```diff
@@ -13,28 +13,34 @@
  */
 package io.trino.plugin.hudi.model;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
```
> The Hudi replace commit metadata model object was also patched to
deserialize Hudi metadata properly. This was needed so that Trino
users (and the tests for the active timeline) do not see duplicated
records for replace commits.

Where's the test about this? This change is unrelated to fix instant file name path, right? Can we separate a PR?

atezs82 replied:
This is tested by the test suite `testReadCopyOnWriteTableWithReplaceCommits` within `TestHudiSparkCompatibility`. If these changes are not applied, clients will see duplicated records before replace commits are cleaned up from the Hudi timeline, so instead of using this assertion in the test:
```java
onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'a1', 1000), (2, 'a2', 2000)");
assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName))
                    .containsOnly(
                            row(1, "a1"),
                            row(2, "a2"));
```
We should use this instead (even though we have inserted only 2 records into the table): 
```
onHudi().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'a1', 1000), (2, 'a2', 2000)");
assertThat(onTrino().executeQuery("SELECT id, name FROM hudi.default." + tableName))
                    .containsOnly(
                            row(1, "a1"),
                            row(2, "a2"),
                            row(1, "a1"),
                            row(2, "a2"));
```
Which looks very strange and I think that these faults are slightly related. Their source is basically next to each other. If you take a look at `HudiTimeline.resetFileGroupsReplaced`:
```java
HudiReplaceCommitMetadata replaceMetadata = HudiReplaceCommitMetadata.fromBytes(
                                metaClient.getActiveTimeline().getInstantDetails(instant).get(),
                                OBJECT_MAPPER,
                                HudiReplaceCommitMetadata.class);
```
- The first fault is that `getInstantDetails` could not create a `Location` instance. That has been fixed in `HudiActiveTimeline`. 
- The second fault is that since the `HudiReplaceCommitMetadata.fromBytes` cannot properly deserialize the class, the list of `partitionToReplaceFileIds` will be empty and clients will see duplicate content.

If you insist, however, we can separate this into two different PRs, that's why I wanted to start the discussion above regarding this in https://github.com/trinodb/trino/pull/18213#issuecomment-1681230801 (see the second paragraph). If that is the case, I will need to change back the tests to this weird form (see above) for the time being (will be fixed in the consecutive PR though). Please let me know what do you think about this.

mosabua commented:
Thank you @ebyhr for all your work on this. 

Also fyi @codope

codope commented:
Thanks a lot @mosabua @ebyhr @atezs82 for helping fix this critical issue!

atezs82 commented:
I also would like to thank @ebyhr for all the work he put into reviewing this change.
